pkg/stash/with_singleflight.go

```diff
@@ -0,0 +1,56 @@
+package stash
+
+import (
+	"sync"
+
+	"github.com/gomods/athens/pkg/config"
+)
+
+// WithSingleflight returns a singleflight stasher.
+// This two clients make two subsequent
+// requests to stash a module, then
+// it will only do it once and give the first
+// response to both the first and the second client.
+func WithSingleflight(s Stasher) Stasher {
+	sf := &withsf{}
+	sf.s = s
+	sf.mp = map[string]struct{}{}
+	sf.subs = map[string][]chan error{}
+
+	return sf
+}
+
+type withsf struct {
+	s Stasher
+
+	mu   sync.Mutex
+	mp   map[string]struct{}
```

Comment:
what is mp?

Reply:
I'll keep mp for short since it's unexported. But I'll add a comment explaining why it's there.

Reply:
it's a way to check if a request is currently in-flight.

Reply:
Actually I can remove it and just use the singleflight map as a way to check an inflight request...I'll update this PR and you can tell me if it's more/less readable and I can revert that commit accordingly.